File: apps/server/internal/model/http.go
package model

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type ErrorClass string

const (
	ErrTimeout       ErrorClass = "timeout"
	ErrHTTP5xx       ErrorClass = "http_5xx"
	ErrRateLimit     ErrorClass = "rate_limit"
	ErrHealthDown    ErrorClass = "health_down"
	ErrMissingKey    ErrorClass = "missing_key"
	ErrNetworkBlock  ErrorClass = "network_blocked"
	ErrAuthError     ErrorClass = "auth_error"
	ErrInputTooLarge ErrorClass = "input_too_large"
	ErrContentSafety ErrorClass = "content_safety"
	ErrUnknown       ErrorClass = "unknown"
)

var fallbackableErrors = map[ErrorClass]bool{
	ErrTimeout: true, ErrHTTP5xx: true, ErrRateLimit: true,
	ErrHealthDown: true, ErrMissingKey: true, ErrNetworkBlock: true,
}

func IsFallbackable(cls ErrorClass) bool { return fallbackableErrors[cls] }

type ProviderError struct {
	Class ErrorClass
	Msg   string
}

func (e *ProviderError) Error() string { return e.Msg }

func newProviderError(cls ErrorClass, msg string) *ProviderError {
	return &ProviderError{Class: cls, Msg: msg}
}

type DeploymentKind string

const (
	DeployPublic  DeploymentKind = "public"
	DeployPrivate DeploymentKind = "private"
)

// NetworkPolicy: "allow_all" | "intranet_only" | "deny_egress" | ""（null）。
type ProviderConnection struct {
	ProviderID     string
	Kind           string // "model" | "visual"
	Name           string
	Protocol       string
	BackendKind    string
	DeploymentKind DeploymentKind
	BaseURL        string
	Credential     string // 已解密，仅用于外发
	Model          string
	TimeoutMs      int
	MaxRetries     int
	NetworkPolicy  string
}

// IsPrivateHost：仅内网/回环目标视为「私有」。比 Node 正则更严（医疗红线：私有/离线 provider 不得出公网）：
// IP 字面量按 CIDR 判定，显式拒绝 169.254/链路本地（云元数据 169.254.169.254）与 0.0.0.0；
// 主机名按锚定 DNS 后缀白名单（避免 10.evil.com / 127.0.0.1.evil.com 前缀绕过）。
func IsPrivateHost(hostname string) bool {
	h := strings.ToLower(strings.TrimSuffix(hostname, "."))
	if h == "localhost" || h == "host.docker.internal" {
		return true
	}
	if ip := net.ParseIP(h); ip != nil {
		if ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast() || ip.IsUnspecified() {
			return false // 拒绝 169.254/fe80（云元数据）与 0.0.0.0/::
		}
		return ip.IsLoopback() || ip.IsPrivate() // 127/8、::1、RFC1918、fc00::/7
	}
	for _, suffix := range []string{".local", ".internal", ".svc"} { // .svc.cluster.local 由 .local 覆盖
		if strings.HasSuffix(h, suffix) {
			return true
		}
	}
	return false
}

// EnforceNetworkPolicy：私有化 provider 命中 deny_egress/intranet_only 时，公网域名目标被拦截。
func EnforceNetworkPolicy(conn ProviderConnection, u *url.URL) error {
	if conn.DeploymentKind != DeployPrivate {
		return nil
	}
	policy := conn.NetworkPolicy
	if policy == "" {
		policy = "intranet_only"
	}
	if policy == "allow_all" {
		return nil
	}
	if !IsPrivateHost(u.Hostname()) {
		return newProviderError(ErrNetworkBlock,
			fmt.Sprintf("私有化 provider「%s」network_policy=%s 禁止出网，但目标 %s 为公网域名，出网网关已拦截", conn.Name, policy, u.Hostname()))
	}
	return nil
}

func resolveURL(baseURL, path string) (*url.URL, error) {
	b := baseURL
	if !strings.HasSuffix(b, "/") {
		b += "/"
	}
	base, err := url.Parse(b)
	if err != nil {
		return nil, err
	}
	ref, err := url.Parse(path)
	if err != nil {
		return nil, err
	}
	return base.ResolveReference(ref), nil
}

// ProviderFetch：统一出站 POST，网络策略校验 + 超时 + 状态码→错误类别映射（不含重试）。返回响应 JSON 字节。
func ProviderFetch(conn ProviderConnection, path string, body any, headers map[string]string) ([]byte, error) {
	u, err := resolveURL(conn.BaseURL, path)
	if err != nil {
		return nil, newProviderError(ErrUnknown, fmt.Sprintf("provider「%s」base_url 非法：%s", conn.Name, conn.BaseURL))
	}
	if err := EnforceNetworkPolicy(conn, u); err != nil {
		return nil, err
	}
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(conn.TimeoutMs)*time.Millisecond)
	defer cancel()
	bodyBytes, _ := json.Marshal(body)
	req, _ := http.NewRequestWithContext(ctx, "POST", u.String(), bytes.NewReader(bodyBytes))
	req.Header.Set("content-type", "application/json")
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	// 重定向亦须再过网络策略（防 302 至公网/元数据绕过出网网关）；跨主机跳转剥离鉴权头。
	client := &http.Client{
		CheckRedirect: func(rreq *http.Request, via []*http.Request) error {
			if len(via) >= 10 {
				return newProviderError(ErrHTTP5xx, "重定向过多")
			}
			if err := EnforceNetworkPolicy(conn, rreq.URL); err != nil {
				return err
			}
			if len(via) > 0 && rreq.URL.Host != via[0].URL.Host {
				rreq.Header.Del("Authorization")
				rreq.Header.Del("x-api-key")
			}
			return nil
		},
	}
	res, err := client.Do(req)
	if err != nil {
		if ctx.Err() == context.DeadlineExceeded {
			return nil, newProviderError(ErrTimeout, fmt.Sprintf("provider「%s」请求超时（%dms）", conn.Name, conn.TimeoutMs))
		}
		var pe *ProviderError
		if errors.As(err, &pe) { // 重定向触发的网络策略拦截，保留错误类别
			return nil, pe
		}
		return nil, newProviderError(ErrHTTP5xx, fmt.Sprintf("provider「%s」连接失败：%s", conn.Name, err.Error()))
	}
	defer res.Body.Close()
	rb, _ := io.ReadAll(res.Body)
	switch {
	case res.StatusCode == 401 || res.StatusCode == 403:
		return nil, newProviderError(ErrAuthError, fmt.Sprintf("provider「%s」鉴权失败（HTTP %d）", conn.Name, res.StatusCode))
	case res.StatusCode == 429:
		return nil, newProviderError(ErrRateLimit, fmt.Sprintf("provider「%s」被限流（HTTP 429）", conn.Name))
	case res.StatusCode >= 500:
		return nil, newProviderError(ErrHTTP5xx, fmt.Sprintf("provider「%s」服务端错误（HTTP %d）", conn.Name, res.StatusCode))
	case res.StatusCode >= 400:
		t := string(rb)
		if len(t) > 200 {
			t = t[:200]
		}
		return nil, newProviderError(ErrUnknown, fmt.Sprintf("provider「%s」请求错误（HTTP %d）%s", conn.Name, res.StatusCode, t))
	}
	return rb, nil
}

func authHeaders(conn ProviderConnection) map[string]string {
	if conn.Credential != "" {
		return map[string]string{"authorization": "Bearer " + conn.Credential}
	}
	return map[string]string{}
}

File: apps/api/src/services/model/http.ts
// c03｜provider 出网网关 + HTTP 客户端（design D4 network_policy；D2 错误分类）
// 私有化 provider 标记 deny_egress / intranet_only 时，出站请求经此网关在传输层拦截公网域名。

export type ErrorClass =
  | "timeout"
  | "http_5xx"
  | "rate_limit"
  | "health_down"
  | "missing_key"
  | "network_blocked"
  | "auth_error"
  | "input_too_large"
  | "content_safety"
  | "unknown";

/** 可 fallback（切换下一 provider）的错误类别；其余直接上抛终止链路（D2） */
export const FALLBACKABLE_ERRORS: ErrorClass[] = [
  "timeout",
  "http_5xx",
  "rate_limit",
  "health_down",
  "missing_key",
  "network_blocked",
];

export function isFallbackable(cls: ErrorClass): boolean {
  return FALLBACKABLE_ERRORS.includes(cls);
}

export class ProviderError extends Error {
  errorClass: ErrorClass;
  constructor(errorClass: ErrorClass, message: string) {
    super(message);
    this.name = "ProviderError";
    this.errorClass = errorClass;
  }
}

export type DeploymentKind = "public" | "private";
export type NetworkPolicy = "allow_all" | "intranet_only" | "deny_egress" | null;

export interface ProviderConnection {
  providerId: string;
  kind: "model" | "visual";
  name: string;
  protocol?: string;
  backendKind?: string;
  deploymentKind: DeploymentKind;
  baseUrl: string;
  /** 已解密凭据（仅用于实际外发，绝不回前端） */
  credential: string;
  model: string;
  timeoutMs: number;
  maxRetries: number;
  networkPolicy: NetworkPolicy;
}

const PRIVATE_HOST_RE =
  /^(localhost|127\.|10\.|192\.168\.|172\.(1[6-9]|2\d|3[01])\.|169\.254\.|host\.docker\.internal$|.*\.local$|.*\.internal$|.*\.svc$|.*\.svc\.cluster\.local$)/i;

export function isPrivateHost(hostname: string): boolean {
  const h = hostname.toLowerCase();
  if (h === "host.docker.internal") return true;
  return PRIVATE_HOST_RE.test(h);
}

/** D4：对私有化 provider 强制执行 network_policy；命中「禁止出网」即便误配公网域名也拦截。 */
export function enforceNetworkPolicy(conn: ProviderConnection, url: URL): void {
  if (conn.deploymentKind !== "private") return;
  const policy = conn.networkPolicy ?? "intranet_only";
  if (policy === "allow_all") return;
  // deny_egress / intranet_only：仅允许内网/回环目标
  if (!isPrivateHost(url.hostname)) {
    throw new ProviderError(
      "network_blocked",
      `私有化 provider「${conn.name}」network_policy=${policy} 禁止出网，但目标 ${url.hostname} 为公网域名，出网网关已拦截`,
    );
  }
}

/** 统一出站请求：网络策略校验 + 超时 + 状态码→错误类别映射。不含重试（重试在 router 内按 provider 策略执行）。 */
export async function providerFetch(
  conn: ProviderConnection,
  path: string,
  body: unknown,
  headers: Record<string, string> = {},
): Promise<unknown> {
  let url: URL;
  try {
    url = new URL(path, conn.baseUrl.endsWith("/") ? conn.baseUrl : `${conn.baseUrl}/`);
  } catch {
    throw new ProviderError("unknown", `provider「${conn.name}」base_url 非法：${conn.baseUrl}`);
  }
  enforceNetworkPolicy(conn, url);

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), conn.timeoutMs);
  let res: Response;
  try {
    res = await fetch(url, {
      method: "POST",
      headers: { "content-type": "application/json", ...headers },
      body: JSON.stringify(body),
      signal: controller.signal,
    });
  } catch (e) {
    if ((e as Error).name === "AbortError") {
      throw new ProviderError("timeout", `provider「${conn.name}」请求超时（${conn.timeoutMs}ms）`);
    }
    // 连接错误（DNS/拒绝连接）按可 fallback 处理
    throw new ProviderError("http_5xx", `provider「${conn.name}」连接失败：${(e as Error).message}`);
  } finally {
    clearTimeout(timer);
  }

  if (res.status === 401 || res.status === 403) {
    throw new ProviderError("auth_error", `provider「${conn.name}」鉴权失败（HTTP ${res.status}）`);
  }
  if (res.status === 429) {
    throw new ProviderError("rate_limit", `provider「${conn.name}」被限流（HTTP 429）`);
  }
  if (res.status >= 500) {
    throw new ProviderError("http_5xx", `provider「${conn.name}」服务端错误（HTTP ${res.status}）`);
  }
  if (res.status >= 400) {
    const text = await res.text().catch(() => "");
    throw new ProviderError("unknown", `provider「${conn.name}」请求错误（HTTP ${res.status}）${text.slice(0, 200)}`);
  }
  return res.json();
}
